### PR TITLE
Add F4E2M1FN and F8E8M0FNU types

### PR DESCRIPTION
**Pull request by sergey-kozub**
This PR adds F4E2M1FN primitive type (4-bit float with 2 bits exponent and 1 bit mantissa), F8E8M0FNU primitive type (8-bit float with 8 bits exponent, no mantissa and no sign) and enables loads/stores in the same way S4/U4 type is implemented.

This will enable using microscaling (MX) formats ([RFC](https://github.com/openxla/xla/discussions/18085)), such as MXFP4.

```c
F4E2M1FN
- Exponent bias: 1
- Maximum stored exponent value: 3 (binary 11)
- Maximum unbiased exponent value: 3 - 1 = 2
- Minimum stored exponent value: 1 (binary 01)
- Minimum unbiased exponent value: 1 − 1 = 0
- Has Positive and Negative zero
- Doesn't have infinity
- Doesn't have NaNs

Additional details:
- Zeros (+/-): S.00.0
- Max normal number: S.11.1 = ±2^(2) x (1 + 0.5) = ±6.0
- Min normal number: S.01.0 = ±2^(0) = ±1.0
- Min subnormal number: S.00.1 = ±2^(0) x 0.5 = ±0.5

F8E8M0FNU
- Exponent bias: 127
- Maximum stored exponent value: 254 (binary 1111'1110)
- Maximum unbiased exponent value: 254 - 127 = 127
- Minimum stored exponent value: 0 (binary 0000'0000)
- Minimum unbiased exponent value: 0 − 127 = -127
- Doesn't have zero
- Doesn't have infinity
- NaN is encoded as binary 1111'1111

Additional details:
- Zeros cannot be represented
- Negative values cannot be represented
- Mantissa is always 1
```

Related PRs:
- https://github.com/openxla/stablehlo/pull/2582
- https://github.com/jax-ml/ml_dtypes/pull/181
- https://github.com/llvm/llvm-project/pull/95392
- https://github.com/llvm/llvm-project/pull/108877
- https://github.com/jax-ml/ml_dtypes/pull/166
- https://github.com/llvm/llvm-project/pull/107127
- https://github.com/llvm/llvm-project/pull/111028

The PR is split into multiple commits just to make the review easier, it is possible that some tests could fail if only some (i.e. not all) of these commits are applied.

### CODE REVIEW

**Comment by sergey-kozub** [comment]
Note: clang-format check fails on issues unrelated to this PR. Also, running with clang-format 18.1.8 yields no issues.

**Comment by reedwm** [comment]
Sorry for the delay in reviewing.

You plan on adding the other MX types, right? If so, and convinient for you, could you add them in the same PR? Normally smaller PRs are better, but all these dtypes touch many of the same places in the code, so it's easier to batch review all of them at once. Granted, FP6 might be a bit weird if packed since their size doesn't divide the byte size -- those potentially should be in a separate PR.

But if adding the other MX types in inconvenient for you, happy to review this PR as-is.

**Comment by sergey-kozub** [comment]
> You plan on adding the other MX types, right? If so, and convinient for you, could you add them in the same PR?

I'll add E8M0 implementation to this PR, as you suggest.
The FP6 types are not implemented yet, I'll look into that later (not a priority).

**Comment by sergey-kozub** [comment]
Added the F8E8M0FNU support patch to this PR. Will update the description soon after.

File: xla/service/elemental_ir_emitter.cc
Comment (reedwm) [inline xla/service/elemental_ir_emitter.cc:797]: Like in EmitF16ToF8e, can you add comments that show the pseduocode of what the LLVM op does? E.g. before defining `llvm::Value* as_int8`, you could have the comment

```
as_int8 = as_int16 >> 9
```

Also like in EmitF16ToF8e, having a few explanation comments and constants would help. E.g. instead of the constant 28 below when defining f4_normal, you could have the constant `mantissa_bits_difference << f4_mantissa_bits`.

It might be worth reusing EmitF16ToF8e for F4 as well. You'd have to handle the fact there is no Inf though, as well as modify handle_halfway_points_F16ToF8, so this might not be worth it. Even if not reusing EmitF16ToF8e, it might be worth copying the general structure for consistency, as most of the code there applies to F4 as well.

And similar comments apply to EmitF4e2m1fnToF16, where you can potentially reuse EmitToF16F8e.

File: xla/service/elemental_ir_emitter.cc
Comment (reedwm) [inline xla/service/elemental_ir_emitter.cc:827]: I find it slightly weird how a large F16 number gets turned into the max F4 value of 6, but an Inf F16 number gets turned to -0. I would expect a very big F16 and an Inf F16 get treated in the same way: either both get turned into 6 or both get turned into -0. But I'm fine with this current behavior if you think it's better than the alternative.

@mooskagh, let me know if you have an opinion here.

File: xla/service/elemental_ir_emitter.cc
Comment (reedwm) [inline xla/service/elemental_ir_emitter.cc:879]: 0x400000 corresponds to FP32 value 2*-127, right? Isn't that value representable by F8e8m0fnu?

File: xla/service/elemental_ir_emitter.cc
Comment (reedwm) [inline xla/service/elemental_ir_emitter.cc:882]: Comment what 0x7F400000 is, e.g. `// 1.5 * 2^127`. And same for 0x400000. And similarly in EmitF8e8m0fnuToF32.

File: third_party/tsl/third_party/py/ml_dtypes/e8m0.patch
Comment (reedwm) [inline third_party/tsl/third_party/py/ml_dtypes/e8m0.patch:4]: Do you plan on submitting this patch to ml_dtypes? It might be easier to submit a PR to ml_dtypes before merging this PR.

File: xla/tests/convert_test.cc
Comment (reedwm) [inline xla/tests/convert_test.cc:2151]: I would just disable this particular value for CPUs, and only when using double.

**Review by reedwm — CHANGES_REQUESTED** [review]
Thanks for the PR!

Note I didn't review the MLIR codegen files (those in gpu/fusions/transforms) since I'm not familiar with those. @mooskagh can you find someone to review these files?

**Review by pifon2a — CHANGES_REQUESTED** [review]

File: xla/service/gpu/fusions/transforms/tests/expand_float_ops.mlir
Comment (sergey-kozub) [reply xla/service/gpu/fusions/transforms/tests/expand_float_ops.mlir:1]: Yes, the tests pass for me. Did you just patch those two files without the other PR's changes?

Comment (pifon2a) [reply xla/service/gpu/fusions/transforms/tests/expand_float_ops.mlir:1]: Yes, only these 2. Other files seem unrelated to running this mlir test.

Comment (pifon2a) [reply xla/service/gpu/fusions/transforms/tests/expand_float_ops.mlir:1]: Hm, weird. But nevermind. Just wanted to confirm that they are running somewhere.

File: xla/service/elemental_ir_emitter.cc
Comment (sergey-kozub) [reply xla/service/elemental_ir_emitter.cc:827]: Only NaN value gets turned to -0, the Inf value is converted to 6.0 (convert_test verifies that).

File: xla/service/elemental_ir_emitter.cc
Comment (sergey-kozub) [reply xla/service/elemental_ir_emitter.cc:879]: It is representable by E8M0, but requires a separate branch (select op).

File: third_party/tsl/third_party/py/ml_dtypes/e8m0.patch
Comment (sergey-kozub) [reply third_party/tsl/third_party/py/ml_dtypes/e8m0.patch:4]: jax/ml-dtypes already has these changes: https://github.com/jax-ml/ml_dtypes/commit/47095b1d2d0c2aa382cf926e8ceaa2a13f0904c5
Once that change propagates to the XLA repo, this patch may be removed.

File: xla/literal.h
Comment (sergey-kozub) [reply xla/literal.h]: Removed.

File: xla/literal_comparison_test.cc
Comment (sergey-kozub) [reply xla/literal_comparison_test.cc]: Good catch, thanks!

File: xla/python/ifrt/dtype.proto
Comment (sergey-kozub) [reply xla/python/ifrt/dtype.proto]: Thanks!

File: xla/python/ifrt/dtype_test.cc
Comment (sergey-kozub) [reply xla/python/ifrt/dtype_test.cc]: Indeed, fixed.

File: xla/tests/convert_test.cc
Comment (sergey-kozub) [reply xla/tests/convert_test.cc]: You're right. Updated the "round up" test, and removed the "round down" as the value `0x0.4p0` is already tested.

File: xla/tests/convert_test.cc
Comment (sergey-kozub) [reply xla/tests/convert_test.cc]: Updated the comment.

File: xla/tests/convert_test.cc
Comment (sergey-kozub) [reply xla/tests/convert_test.cc]: Done.

File: xla/tests/convert_test.cc
Comment (sergey-kozub) [reply xla/tests/convert_test.cc:1982]: The reason is the same why all of such FP8 tests are also marked as DISABLED_ON_CPU.
When CPU converts FP32->FP8 it inserts an intermediary conversion of FP32->FP16->FP8, which loses least significant mantissa bits.

The specific failing value in this test is `0x1.000002p-2` (Smallest number that doesn't underflow) which is slightly higher than 0.25.
Converting this value directly FP32->FP8 would correctly yield 0.5 (due to how round-to-even algorithm works).
However, when first converted to FP16, the value becomes 0.25 exactly, which then rounds down to zero.

File: xla/primitive_util.h
Comment (sergey-kozub) [reply xla/primitive_util.h:413]: The `F8E8M0FNU` type is both an FP8 (as it is 8 bits) and a special MX type. So adding it to either set is ok (as I see it).
This type is also very different from the other FP8 types (no mantissa, no sign). If we add it to `IsF8Type`, then we'd have to verify all of its usages (there are many in the code). Classifying it as `IsMXType` will avoid adding a special case for it in many places (i.e. writing: IsF8Type(x) && x != F8E8M0FNU)

File: xla/util.h
Comment (sergey-kozub) [reply xla/util.h:663]: `Eigen::numext::signbit` returns either 0x00 for positive input or 0xFF for negative input, not a single bit.
I updated the code here to account for this fact (i.e. removed a select op below). I also verified that it works correctly for FP4.

File: xla/service/gpu/fusions/transforms/lower_tensors.cc
Comment (sergey-kozub) [reply xla/service/gpu/fusions/transforms/lower_tensors.cc:1]: Added "f4_constant", "transfer_read_f4" and "transfer_write_f4".
BTW, there's no "transfer_read_i4" in that file.

**Comment by sergey-kozub** [comment]
I believe I addressed all the prior comments. PTAL when possible.

**Review by pifon2a — APPROVED** [review]
Thank you!

**Review by reedwm — APPROVED** [review]

**Comment by mooskagh** [comment]
I'm trying to merge this PR.
It references `float8.h` in `ml_dtypes.h`, but I cannot find it either in this repository, nor internally (it was deleted a year ago).

**Comment by nouiz** [comment]
What is the status of this PR?

**Comment by mooskagh** [comment]
The internal TPU part of the XLA requires corresponding changes to be able to integrate this PR, so it's not just "merging" but rather implementing -- that's why it's taking time.

However, meanwhile the PR is getting out of sync and it's getting harder to rebase it. Could you resolve the conflicts so that the PR gets rebased?

Otherwise, there are also these changes to be made:

- In `mxfloat.h`, the patching of Eigen library through defining template function specialization doesn't go well with header modules. Could you replace it with function override? (remove the `template <> \` line)
- `e8m0.patch` is not needed as the changes are already intergrated upstream.
- In `primitive_util_test.cc`, the test function gets over the limit of 800 lines, add `// NOLINT(readability/fn_size)` in the end (I'll fix it properly afterwards)
- In tests, instead of `#ifdef XLA_TEST_BACKEND_CPU`, use `if (client_->platform()->id() == stream_executor::host::kHostPlatformId)`
- `pjrt_c_api.h` needs a version increment, and a corresponding entry in `pjrt/c/CHANGELOG.md`
- Some tests fail, e.g. `triton_support_test.cc:573` fails with `value 0 is out of range for literal's primitive type F8E8M0FNU namely [5.87747e-39, 1.70141e+38].`

**Comment by sergey-kozub** [comment]
I'll work on rebasing this PR

**Comment by sergey-kozub** [comment]
> Could you resolve the conflicts so that the PR gets rebased?

I rebased this PR and addressed your comments, PTAL.

I haven't touched the `mxfloat.h` file, as I don't see what it breaks. If it breaks some internal tests, please create a patch.

**Review by reedwm — APPROVED** [review]

**Comment by reedwm** [comment]
I'm working on merging this, taking over from @mooskagh who is OOO. No need to address merge conflicts, I am resolving internally.

**Comment by sergey-kozub** [comment]
> This PR was rolled back in [e19d979](https://github.com/openxla/xla/commit/e19d979eb102c4ee6854b057ed02e23567275582)!

What was the reason for the rollback? @jaeyoo 
If you could give me the details, I could help resolve the issue.

**Comment by reedwm** [comment]
This broke the TensorFlow Android build. The error is

```
ERROR: /[tmpfs/bazel_output/_bazel_kbuilder/f2d52ca1f092ccbe254cc98c3dc90790/external/local_tsl/tsl/platform/BUILD:1327](https://cs.corp.google.com/piper///depot/google3/tmpfs/bazel_output/_bazel_kbuilder/f2d52ca1f092ccbe254cc98c3dc90790/external/local_tsl/tsl/platform/BUILD?l=1327&cl=708401281):11: Compiling tsl/platform/abi.cc failed: (Exit 1): clang failed: error executing command (from target @local_tsl//tsl/platform:abi) external/androidndk/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/clang -gcc-toolchain external/androidndk/ndk/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64 -target ... (remaining 80 arguments skipped)
In file included from [external/local_tsl/tsl/platform/abi.cc:16](https://cs.corp.google.com/piper///depot/google3/external/local_tsl/tsl/platform/abi.cc?l=16&cl=708401281):
In file included from [external/local_tsl/tsl/platform/abi.h:21](https://cs.corp.google.com/piper///depot/google3/external/local_tsl/tsl/platform/abi.h?l=21&cl=708401281):
In file included from [external/local_tsl/tsl/platform/types.h:19](https://cs.corp.google.com/piper///depot/google3/external/local_tsl/tsl/platform/types.h?l=19&cl=708401281):
In file included from [external/local_xla/xla/tsl/platform/types.h:22](https://cs.corp.google.com/piper///depot/google3/external/local_xla/xla/tsl/platform/types.h?l=22&cl=708401281):
In file included from [external/local_tsl/tsl/platform/ml_dtypes.h:19](https://cs.corp.google.com/piper///depot/google3/external/local_tsl/tsl/platform/ml_dtypes.h?l=19&cl=708401281):
[external/ml_dtypes/include/float8.h:117](https://cs.corp.google.com/piper///depot/google3/external/ml_dtypes/include/float8.h?l=117&cl=708401281):12: error: calling a protected constructor of class 'ml_dtypes::mxfloat_internal::float6_e2m3fn'
    return Derived(rep, ConstructFromRepTag{});
           ^
[bazel-out/arm64-v8a-opt/bin/external/ml_dtypes/_virtual_includes/mxfloat/ml_dtypes/include/mxfloat.h:152](https://cs.corp.google.com/piper///depot/google3/bazel-out/arm64-v8a-opt/bin/external/ml_dtypes/_virtual_includes/mxfloat/ml_dtypes/include/mxfloat.h?l=152&cl=708401281):27: note: in instantiation of member function 'ml_dtypes::float8_internal::float8_base<ml_dtypes::mxfloat_internal::float6_e2m3fn>::FromRep' requested here
    return float6_e2m3fn::FromRep(0b0'01'000);
                          ^
[bazel-out/arm64-v8a-opt/bin/external/ml_dtypes/_virtual_includes/mxfloat/ml_dtypes/include/mxfloat.h:77](https://cs.corp.google.com/piper///depot/google3/bazel-out/arm64-v8a-opt/bin/external/ml_dtypes/_virtual_includes/mxfloat/ml_dtypes/include/mxfloat.h?l=77&cl=708401281):15: note: declared protected here
  using Base::Base;
              ^
```

I'm trying to figure out how to reproduce this on a local client where I can modify code, then will try to figure out how to fix it. I don't think you can reproduce this, but let me know if you have any ideas what the issue could be.

The protected constructor the error is complaining about is called in `float8_base` [here](https://github.com/jax-ml/ml_dtypes/blob/560866cfbb44990c9c68b82e878bf8ee7972a595/ml_dtypes/include/float8.h#L116-L118). The constructor is for a subclass of `float8_base`, `float6_e2m3fn`, but is defined in `float8_base` and brought into the subclass with a using declaration: `using Base::Base`. I'm not enough of a C++ wizard to know if this is legal, but all compilers except the one in the Android build are accepting this.

**Comment by sergey-kozub** [comment]
Do you know how to reproduce the compilation failure in OSS?
What is the exact compiler and its version that fails?
I think we could create an XLA patch that will mitigate the issue and then properly upstream it to JAX ml_dtypes.

I'd really like to get this PR pushed ASAP, as it blocks the block scaled dot feature in XLA (on the recently released Blackwell).

**Comment by reedwm** [comment]
Sorry for the delays here. I'm still working on reproducing the issue with custom changes to ml_dtypes, which is the only way to debug this issue. But I think I'll be able to reproduce soon, at which point I should be able to fix soon as well.

The failing Kokoro build is compiling TensorFlow through [this script](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/presubmit/ubuntu_16/android/build.sh). I think the issue is the build is using a fairly old version of the Android NDK, `android-ndk-r21e`, as a different internal build passes with this PR with r25. This different build differs in other ways as well, but none seem like they would affect whether the build passes.

I think ml_dtypes does have valid C++, and that the Android compiler producing the error is incorrect. I tried creating a reproducer at https://godbolt.org/z/qcPPxavn9 that has roughly the same pattern of inheriting a protected constructor as ml_dtypes did, but it passed on all the godbolt compilers I've tried. I'm not sure what compiler Android NDK uses, and am unsure why it's failing to build ml_dtypes.

**Comment by reedwm** [comment]
I found the Android NDK r21e uses clang 9.0.9. I found a small program to reproduce the issue:

```c++
template <typename T>
class A {
protected:
  A(int, int) {}
public:
  static T f() {
    return T(1, 2);
  }
};

class B : public A<B> {
  using A<B>::A;
};

int main() {
  B::f();
}
```

Seems it breaks only when the constructor has two or more arguments, which is strange. Clang 13 fixes the issue. I wonder what the issue in clang was.

**Comment by reedwm** [comment]
I created a fix in https://github.com/jax-ml/ml_dtypes/commit/32059235df380c51ed821adbb3f75df78de4c346 for Clang 12 and below. I also have another pending fix in https://github.com/jax-ml/ml_dtypes/pull/254 that resolves a different internal issue. I originally put this fix in the same CL that merged your PR, which showed up as https://github.com/jax-ml/ml_dtypes/commit/986e373f20923def7720abbd91c15a53557a7286, but it was reverted when the PR was rolled back.

Once the second ml_dtypes PR is in, I'll create a separate change to update the JAX and TF ml_dtypes version to the latest commit, which should unblock this PR. Can you create a new copy of this PR to merge? Make sure it has the contents of the merge commit https://github.com/openxla/xla/commit/2533c35067b7806aef1f08eb0bd16391a568344d, and not the contents of the original PR, since I made changes to fix internal tests when merging it. Alternatively, if you prefer, I can directly rollforward the PR as a rollback of a rollback.

**Comment by sergey-kozub** [comment]
Thanks, Reed! I'll proceed with rebasing https://github.com/openxla/xla/commit/2533c35067b7806aef1f08eb0bd16391a568344d to HEAD and create a new PR.

**Comment by reedwm** [comment]
I updated the ml_dtypes version used by TF in https://github.com/tensorflow/tensorflow/commit/42bd1d3c6554573a8619338fcf0115d5a56bce9e, and the TensorFlow Android build is still passing. So we should be able to merge the PR smoothly this time once you create the new PR :)

**Comment by sergey-kozub** [comment]
Created a new PR https://github.com/openxla/xla/pull/21380 (rebased)

**Pull request closed without merging**
